python/lsst/daf/butler/datastores/webdavDatastore.py

```diff
@@ -0,0 +1,271 @@
+
+    def _read_artifact_into_memory(self, getInfo: DatastoreFileGetInformation,
+                                   ref: DatasetRef, isComponent: bool = False) -> Any:
+        location = getInfo.location
+
+        response = self.session.get(location.uri.geturl())
+        if response.status_code != 200:
+            errorcode = response.status_code
+            if errorcode == 403:
+                raise FileNotFoundError(f"Dataset with Id {ref.id} not accessible at "
+                                        f"expected location {location.uri}. Forbidden "
+                                        "operation error occured. Verify permissions are granted for "
+                                        "your user and that file exists. ")
+            if errorcode == 404:
+                errmsg = f"Dataset with Id {ref.id} does not exists at expected location \
+                         {location.uri.geturl()}."
+                raise FileNotFoundError(errmsg)
+            raise FileNotFoundError(f"There was an error getting file at {location.uri.geturl()}, \
+                                    status code : {errorcode}")
+
+        log.debug("Successful read on file %s", location.uri.geturl())
+        storedFileInfo = getInfo.info
+        if len(response.content) != storedFileInfo.file_size:
+            raise RuntimeError("Integrity failure in Datastore. Size of file {} ({}) does not"
+                               " match recorded size of {}".format(location.path, len(response.content),
+                                                                   storedFileInfo.file_size))
+
+        serializedDataset = response.content
+
+        formatter = getInfo.formatter
+        try:
+            result = formatter.fromBytes(serializedDataset,
+                                         component=getInfo.component if isComponent else None)
+        except NotImplementedError:
+            log.debug("Formatter does not have extension, using temporary file.")
+            tmpLoc = Location(".", "temp")
+            tmpLoc = formatter.makeUpdatedLocation(tmpLoc)
+            with tempfile.NamedTemporaryFile(suffix=tmpLoc.getExtension()) as tmpFile:
+                tmpFile.write(serializedDataset)
+                tmpFile.flush()
+                formatter._fileDescriptor.location = Location(*os.path.split(tmpFile.name))
+                result = formatter.read(component=getInfo.component if isComponent else None)
+        except Exception as e:
+            raise ValueError(f"Failure from formatter '{formatter.name()}' for dataset {ref.id}"
+                             f" ({ref.datasetType.name} from {location.uri}): {e}") from e
+
+        return self._post_process_get(result, getInfo.readStorageClass, getInfo.assemblerParams,
+                                      isComponent=isComponent)
+
+    def _write_in_memory_to_artifact(self, inMemoryDataset: Any, ref: DatasetRef) -> StoredFileInfo:
```

Comment:
It's clear that this needs to be rewritten using ButlerURI to do all the work so that it would be generic -- but not on this pull request.